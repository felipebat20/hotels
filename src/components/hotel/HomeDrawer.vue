<template>
  <q-card-section class="q-pb-none">
    <div class="row justify-between full-width items-center">
      <div class="q-ma-none q-mb-sm row items-center q-gutter-x-sm">
        {{ rating.toFixed(1) }}

        <q-rating
          v-model="rating"
          readonly
          size="18px"
          :max="5"
          color="yellow-7"
        />
      </div>
    </div>

    <p class="text-h4 q-mb-none two-lines-truncate">
      {{ hotel.name }}
    </p>

    <p class="text-grey-6 q-mb-none">
      {{ hotel.address.fullAddress }}
    </p>
  </q-card-section>

  <q-card-section>
    <q-carousel
      v-model="slide"
      :autoplay="9999"
      animated
      arrows
      class="no-scroll"
      control-color="white"
      control-text-color="grey"
      control-type="regular"
      height="512px"
      infinite
      navigation-icon="radio_button_unchecked"
    >
      <q-carousel-slide
        v-for="(image, index) in hotel.images"
        :key="index"
        :name="index"
        class="q-pa-none"
        style="height: 512px"
      >
        <q-img
          :src="image"
          spinner-color="primary"
          fit="cover"
          height="512px"
          class="full-width"
          no-native-menu
        />
      </q-carousel-slide>

      <template v-slot:control>
        <div class="image-counter text-white q-px-md q-py-xs">
          {{ slide + 1 }}/{{ hotel.images.length }}
        </div>
      </template>
    </q-carousel>
  </q-card-section>

  <q-card-section v-if="hotel.amenities.length">
    <p class="text-h5">
      Facilidades do hotel
    </p>

    <q-separator />

    <div class="q-my-lg grid">
      <span
        v-for="amenity in getAmenities"
        class="text-grey-6 row justify-start items-center q-pb-sm"
        :key="amenity.key"
      >
        <q-icon
          :name="amenity.icon"
          size="sm"
          class="q-mr-xs"
        />

        {{ amenity.label }}
      </span>
    </div>

    <div
      v-if="hotel.amenities.length > 4"
      class="row justify-center q-mb-lg"
    >
      <q-btn
        rounded
        no-caps
        color="primary"
        outline
        @click="toggleFacilities"
      >
        {{ show_facilities ? 'Mostrar menos facilidades' : 'Mostrar todas as facilidades' }}
      </q-btn>
    </div>

    <q-separator />
  </q-card-section>

  <q-card-section v-if="hotel.description">
    <p class="text-h5">
      Conhe√ßa um pouco mais
    </p>

    <p class="text-grey-6">
      {{ hotel.description }}
    </p>
  </q-card-section>
</template>

<script setup lang="ts">
import { ref, computed } from 'vue';

import type { HotelInterface } from 'src/interfaces/HotelInterface';

const { hotel } = defineProps<{
  hotel: HotelInterface;
}>();

const rating = ref(+hotel.stars);
const slide = ref(0);
const show_facilities = ref(false);
const amenities_icons = {
  AIR_CONDITIONING: 'ac_unit',
  AUDITORIUM: 'vertical_shades_closed',
  BREAKFAST: 'breakfast_dining',
  FITNESS_CENTER: 'fitness_center',
  LAUNDRY: 'local_laundry_service',
  MASSAGE: 'spa',
  MEETING_ROOM: 'meeting_room',
  PARKING: 'local_parking',
  PETS: 'pets',
  PLAY_GROUND: 'mood',
  POOL: 'pool',
  PUB: 'sports_bar',
  RECEPTION_24_HOURS: 'timer',
  RESTAURANT: 'restaurant',
  ROOM_SERVICE: 'room_service',
  ROOM_TV: 'tv',
  SAFE: 'lock',
  SPA: 'self_improvement',
  STAGE: 'vertical_shades',
  STEAM_ROOM: 'hot_tub',
  TRANSFER: 'cases',
  WI_FI: 'wifi',
}

const getAmenities = computed(() => {
  return hotel.amenities.map((amenity) => {
    return {
      key: amenity.key,
      label: amenity.label,
      icon: amenities_icons[amenity.key],
    };
  }).filter((_, index) => {
    return index < 4 || show_facilities.value;
  });
});

const toggleFacilities = () => {
  show_facilities.value = ! show_facilities.value;
}
</script>


<style scoped>
.grid {
  display: grid;
  grid-template-columns: repeat(4, auto);
  justify-content: space-between;
}
</style>