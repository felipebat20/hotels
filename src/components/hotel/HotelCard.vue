<template>
  <q-card class="row no-shadow q-mb-md hotel-card-container">
    <q-card-section class="image-container col-12 col-sm-3 q-pa-none">
      <q-carousel
        v-model="slide"
        infinite
        animated
        arrows
        :autoplay="9999"
        height="250px"
        navigation-icon="radio_button_unchecked"
        control-type="regular"
        control-color="white"
        control-text-color="grey"
        class="custom-border"
      >
        <q-carousel-slide
          v-for="(image, index) in hotel.images"
          :key="image"
          :name="index"
          :img-src="image"
          class="fit"
        />
      </q-carousel>
    </q-card-section>

    <q-card-section class="col-12 col-sm-6">
      <p
        class="text-h5 q-mb-none cursor-pointer"
        @click="handleHotelClick"
      >
        {{ hotel.name }}
      </p>

      <span class="text-grey-6">
        {{ hotel.address.fullAddress }}
      </span>

      <div class="q-ma-none q-mb-sm row items-center q-gutter-x-sm">
        {{ hotel_rating }}

        <q-rating
          v-model="hotel_rating"
          readonly
          size="18px"
          :max="5"
          color="yellow"
        />

        <template v-if="hotel.amenities?.length">
          <q-separator vertical color="grey" />

          <q-icon
            v-for="amenity in getParsedHotelAmenities"
            :key="amenity.key"
            :name="amenity.icon"
            size="xs"
          >
            <q-tooltip class="bg-primary text-bold">
              {{ amenity.label }}
            </q-tooltip>
          </q-icon>
        </template>
      </div>

      <q-chip
        v-if="canRefund"
        color="grey"
        text-color="white"
        class="text-weight-medium"
        square
      >
        Reembols√°vel
      </q-chip>
    </q-card-section>

    <q-card-section class="col-12 col-sm-3">
      <q-separator vertical color="grey-3" class="custom-separator" />

      <p class="q-mb-xs text-grey-6 text-caption">
        A partir de:
      </p>

      <p class="q-mb-xs">
        R$ <span class="text-h5">
          {{ getHotelPrice }}
        </span>
      </p>

      <p class="q-mb-none text-grey-6 text-caption">
        R${{ getPerNightCurrency }}/noite
      </p>

      <p class="text-grey-7 q-mb-xl text-caption">
        Impostos inclusos
      </p>

      <div class="row">
        <q-btn
          rounded
          no-caps
          color="primary"
          size="md"
          class="col-12 col-sm-auto"
          style="min-width: 140px"
          @click="handleHotelClick"
        >
          Selecionar
        </q-btn>
      </div>
    </q-card-section>
  </q-card>
</template>

<script setup lang="ts">
import { ref, computed } from 'vue';

import type { HotelInterface } from '@/interfaces/HotelInterface';
import type { Amenity } from '@/interfaces/AmenityInterface';

import { formatCurrency } from '@/utils/format-currency';

const slide = ref(0);
const { hotel } = defineProps<{
  hotel: HotelInterface;
}>();

const emit = defineEmits(['set-selected-hotel']);

const handleHotelClick = () => {
  emit('set-selected-hotel', hotel);
};

const hotel_rating = ref(+hotel.stars);

const amenities_icons = {
  AIR_CONDITIONING: 'ac_unit',
  AUDITORIUM: 'vertical_shades_closed',
  BREAKFAST: 'breakfast_dining',
  FITNESS_CENTER: 'fitness_center',
  LAUNDRY: 'local_laundry_service',
  MASSAGE: 'spa',
  MEETING_ROOM: 'meeting_room',
  PARKING: 'local_parking',
  PETS: 'pets',
  PLAY_GROUND: 'mood',
  POOL: 'pool',
  PUB: 'sports_bar',
  RECEPTION_24_HOURS: 'timer',
  RESTAURANT: 'restaurant',
  ROOM_SERVICE: 'room_service',
  ROOM_TV: 'tv',
  SAFE: 'lock',
  SPA: 'self_improvement',
  STAGE: 'vertical_shades',
  STEAM_ROOM: 'hot_tub',
  TRANSFER: 'cases',
  WI_FI: 'wifi',
}

const getParsedHotelAmenities = computed(() => {
  return hotel.amenities?.map((amenity: Amenity) => ({
    key: amenity.key,
    label: amenity.label,
    icon: amenities_icons[amenity.key],
  }));
});

const canRefund = computed(() => {
  return hotel.hasRefundableRoom && ! hotel.nonRefundable;
});

const getHotelPrice = computed(() => {
  return formatCurrency(hotel.price);
});

const getPerNightCurrency = computed(() => {
  return formatCurrency(hotel.price / 3);
});

</script>

<style lang="scss" scoped>
.fit { object-fit: cover; }
.image-container {
  :deep(.q-carousel) {
    .q-panel.scroll { overflow: hidden !important; }
    .q-carousel__prev-arrow--horizontal { left: .5rem; }
    .q-carousel__next-arrow--horizontal { right: .5rem; }
    .q-carousel__arrow {
      button {
        width: 1.125rem;
        height: 1.125rem;
        min-width: unset;
        min-height: unset;

        i { font-size: 1rem; }
      }
    }
  }
}

.price-container { border-left: 1px solid $grey-3!important; }
.custom-border {
  border-top-left-radius: 10px;
  border-bottom-left-radius: 10px;
}

.hotel-card-container {
  border-radius: 20px;
  overflow: hidden;
}

.custom-separator {
  height: 100%;
  position: absolute;
  left: 0;
  top: 0;
}
</style>
