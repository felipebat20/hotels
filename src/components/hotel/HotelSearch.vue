<template>
  <q-card class="bg-white rounded-borders-md">
    <q-card-section class="q-px-md q-py-sm">
      <h2 class="text-h5 q-ma-none">Reservar hotel</h2>
    </q-card-section>

    <q-separator />

    <q-card-section class="column q-gutter-md">
      <div>
        <label>
          <span>
            Destino
          </span>

          <q-select
            v-model="city_search"
            :options="filteredPlaces"
            use-input
            hide-selected
            fill-input
            dense
            outlined
            hide-dropdown-icon
            input-debounce="0"
            placeholder="Cidade, UF"
            option-value="placeId"
            class="place-select"
            @filter="handleFilterPlaces"
          />
        </label>
      </div>

      <div class="row justify-end">
        <q-btn
          rounded
          no-caps
          color="primary"
          size="md"
          class="col-12 col-sm-auto"
          style="min-width: 180px"
        >
          Buscar

          <template v-slot:loading>
            <q-spinner-hourglass class="on-left" />

            Buscando...
          </template>
        </q-btn>
      </div>
    </q-card-section>
  </q-card>
</template>

<script setup lang="ts">
import { ref, computed } from 'vue';

import places from '@/data/place.json';

const city_search = ref('');

const options = ref(places);

const filteredPlaces = ref();

const getFormattedPlaces = computed(() => {
  return options.value.map((city) => {
    return {
      label: `${city.name}, ${city.state.name}`,
      value: city.placeId,
      ...city
    };
  });
});


const handleFilterPlaces = (value: string, update: (callbackFn: () => void) => void) => {
  update(() => {
    const new_search = value.toLowerCase();

    filteredPlaces.value = getFormattedPlaces.value.filter((place) => {
      return place.label.toLowerCase().includes(new_search) || place.state.name.toLowerCase().includes(new_search);
    })
  });
};
</script>
