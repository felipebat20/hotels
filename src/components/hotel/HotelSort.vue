<template>
  <div class="row items-center">
    <p class="q-mb-none text-grey-7">
      Organizar por
    </p>

    <q-select
      v-model="selected_sort"
      :options="sortOptions"
      dense
      borderless
      options-dense
      @update:model-value="handleSort"
    >
      <template v-slot:selected>
        <span class="q-mb-none q-ml-xs text-blue text-weight-medium text-italic">
          {{ selected_sort.label }}
        </span>
      </template>
    </q-select>
  </div>
</template>

<script setup lang="ts">
import { ref } from 'vue';
import type { Ref } from 'vue';

import { useHotelsStore } from 'src/stores/useHotelsStore';

interface SortOption {
  label: string;
  value: string;
}

const emit = defineEmits(['fetch-hotels'])

const sortOptions: Ref<SortOption[]> = ref([
  {
    label: 'Recomendados',
    value: 'recommended'
  },

  {
    label: 'Melhor avaliados',
    value: 'best_rated'
  },
]);

const selected_sort: Ref<SortOption> = ref(sortOptions.value[0] as SortOption);
const hotelsStore = useHotelsStore();

const handleSort = async ({ value }: { value: string }) => {
  hotelsStore.control_flow.sortBy = value;
  hotelsStore.hotels = [];
  hotelsStore.control_flow.request_pending = true;

  await hotelsStore.fetchHotels({
    sortBy: value,
    name: hotelsStore.control_flow.name,
    placeId: hotelsStore.control_flow.placeId,
    page: 1,
  });

  hotelsStore.control_flow.request_pending = false;
};
</script>