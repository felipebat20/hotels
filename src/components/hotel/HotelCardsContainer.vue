<template>
  <div
    v-if="hotelsStore.control_flow.request_pending && ! hotelsStore.hotels.length"
    class="row justify-center q-my-md"
  >
    <q-spinner-dots color="primary" size="40px" />
  </div>

  <template v-else>
    <q-infinite-scroll
      @load="handleLoad"
      :offset="200"
      :disable="! hasMoreHotels"
    >
      <template v-if="hotelsStore.hotels.length">
        <HotelCard
          v-for="hotel in hotelsStore.hotels"
          :key="hotel.id"
          :hotel="hotel"
          @set-selected-hotel="handleHotelClick"
        />
      </template>


      <template v-else>
        <div class="column items-center q-pa-xl">
          <p class="text-h6 text-grey-6">
            Nenhum hotel encontrado
          </p>
        </div>
      </template>

      <template #loading>
        <div class="row justify-center q-my-md">
          <q-spinner-dots
            color="primary"
            size="40px"
          />
        </div>
      </template>
    </q-infinite-scroll>

    <div
      v-if="! hotelsStore.control_flow.has_next_page && hotelsStore.hotels.length"
      class="row justify-center"
    >
      <p class="text-grey-6">
        Não existem mais hotéis para serem exibidos
      </p>
    </div>
  </template>
</template>

<script setup lang="ts">
import { ref, computed } from 'vue';

import HotelCard from 'src/components/hotel/HotelCard.vue';

import { HotelInterface } from 'src/interfaces/HotelInterface';

import { useHotelsStore } from 'src/stores/useHotelsStore';

const hotelsStore = useHotelsStore();
const request_pending = ref(false);


const hasMoreHotels = computed(() => {
  return hotelsStore.control_flow.has_next_page;
});

const handleLoad = async (index: number, done: () => void) => {
  hotelsStore.control_flow.request_pending = 1 === index;
  request_pending.value = true;

  await hotelsStore.fetchHotels({
    sortBy: hotelsStore.control_flow.sortBy,
    name: hotelsStore.control_flow.name,
    placeId: hotelsStore.control_flow.placeId,
    page: index,
  });

  request_pending.value = false;

  done();
};

const emit = defineEmits(['set-selected-hotel']);
const handleHotelClick = (hotel: HotelInterface) => {
  emit('set-selected-hotel', hotel);
};
</script>
