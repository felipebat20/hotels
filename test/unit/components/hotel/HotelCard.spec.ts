import { installQuasarPlugin } from '@quasar/quasar-app-extension-testing-unit-vitest';
import { mount } from '@vue/test-utils';
import { describe, expect, it } from 'vitest';

import HotelCard from '@/components/hotel/HotelCard.vue';

import type { HotelInterface } from 'src/interfaces/HotelInterface';

import hotel from '@/data/hotel.json';

installQuasarPlugin();

describe('HotelCard', () => {
  const createWrapper = ({ props = {} }: { props: { hotel?: HotelInterface } } = { props: {} }) => {
    return mount(HotelCard, {
      props: {
        hotel: hotel[0]?.hotels[0],
        ...props,
      },
    });
  };

  it('should render component properly', () => {
    const wrapper = createWrapper();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display hotel name correctly', () => {
    const wrapper = createWrapper();
    const hotelName = wrapper.find('.text-h5');

    expect(hotelName.text()).toBe(hotel[0]?.hotels[0]?.name);
  });

  it('should display hotel address correctly', () => {
    const wrapper = createWrapper();
    const address = wrapper.find('.text-grey-6');

    expect(address.text()).toBe(hotel[0]?.hotels[0]?.address.fullAddress);
  });

  it('should display correct hotel price', () => {
    const wrapper = createWrapper();
    const price = wrapper.findAll('.text-h5').at(1);

    expect(price?.text()).toBe(wrapper.vm.getHotelPrice);
  });

  it('should emit event when select button is clicked', async () => {
    const wrapper = createWrapper();
    await wrapper.findAll('button').at(2)?.trigger('click');

    expect(wrapper.emitted('set-selected-hotel')).toBeTruthy();
  });
});
