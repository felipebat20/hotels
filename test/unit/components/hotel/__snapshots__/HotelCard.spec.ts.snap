// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HotelCard > should render component properly 1`] = `
"<div data-v-be666f43="" class="q-card row no-shadow q-mb-md hotel-card-container">
  <div data-v-be666f43="" class="q-card__section q-card__section--vert image-container col-12 col-sm-3 q-pa-none">
    <div data-v-be666f43="" class="q-carousel q-panel-parent q-carousel--without-padding q-carousel--arrows-horizontal custom-border" style="height: 250px;">
      <div class="q-carousel__slides-container">
        <transition-stub appear="false" persisted="false" css="true">
          <div class="q-panel scroll" style="--q-transition-duration: 300ms;" role="tabpanel">
            <div data-v-be666f43="" class="q-carousel__slide fit" style="background-image: url(&quot;https://s3.amazonaws.com/e-htl/uploads/hotels/10415/10415_10.JPG&quot;);"></div>
          </div>
        </transition-stub>
      </div>
      <div class="q-carousel__control q-carousel__arrow q-carousel__prev-arrow q-carousel__prev-arrow--horizontal absolute flex flex-center"><button class="q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--round bg-white text-grey q-btn--actionable q-focusable q-hoverable q-btn--dense" tabindex="0" type="button" regular="true"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><i class="q-icon notranslate material-icons" aria-hidden="true" role="img">chevron_left</i></span></button></div>
      <div class="q-carousel__control q-carousel__arrow q-carousel__next-arrow q-carousel__next-arrow--horizontal absolute flex flex-center"><button class="q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--round bg-white text-grey q-btn--actionable q-focusable q-hoverable q-btn--dense" tabindex="0" type="button" regular="true"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><i class="q-icon notranslate material-icons" aria-hidden="true" role="img">chevron_right</i></span></button></div>
    </div>
  </div>
  <div data-v-be666f43="" class="q-card__section q-card__section--vert col-12 col-sm-6">
    <p data-v-be666f43="" class="text-h5 q-mb-none cursor-pointer">Nacional Inn Belo Horizonte</p><span data-v-be666f43="" class="text-grey-6">Rua Espírito Santo, 215 - Centro</span>
    <div data-v-be666f43="" class="q-ma-none q-my-sm row items-center q-gutter-x-sm">3.0 <div data-v-be666f43="" class="q-rating row inline items-center q-rating--non-editable text-yellow-7 no-shadow" style="font-size: 18px;" role="radiogroup" aria-readonly="true">
        <div class="q-rating__icon-container flex flex-center" role="radio" aria-checked="false" aria-label="grade 1"><i class="q-icon notranslate material-icons q-rating__icon q-rating__icon--active text-yellow-7" aria-hidden="true">grade</i></div>
        <div class="q-rating__icon-container flex flex-center" role="radio" aria-checked="false" aria-label="grade 2"><i class="q-icon notranslate material-icons q-rating__icon q-rating__icon--active text-yellow-7" aria-hidden="true">grade</i></div>
        <div class="q-rating__icon-container flex flex-center" role="radio" aria-checked="true" aria-label="grade 3"><i class="q-icon notranslate material-icons q-rating__icon q-rating__icon--active text-yellow-7" aria-hidden="true">grade</i></div>
        <div class="q-rating__icon-container flex flex-center" role="radio" aria-checked="false" aria-label="grade 4"><i class="q-icon notranslate material-icons q-rating__icon text-yellow-7" aria-hidden="true">grade</i></div>
        <div class="q-rating__icon-container flex flex-center" role="radio" aria-checked="false" aria-label="grade 5"><i class="q-icon notranslate material-icons q-rating__icon text-yellow-7" aria-hidden="true">grade</i></div>
      </div>
      <hr data-v-be666f43="" class="q-separator q-separator--vertical bg-grey" aria-orientation="vertical"><i data-v-be666f43="" class="q-icon notranslate material-icons" style="font-size: 18px;" aria-hidden="true">wifi
        <!---->
      </i><i data-v-be666f43="" class="q-icon notranslate material-icons" style="font-size: 18px;" aria-hidden="true">restaurant
        <!---->
      </i><i data-v-be666f43="" class="q-icon notranslate material-icons" style="font-size: 18px;" aria-hidden="true">room_service
        <!---->
      </i>
    </div>
    <div data-v-be666f43="" class="q-chip row inline no-wrap items-center bg-grey text-white q-chip--colored q-chip--square text-weight-medium">
      <div class="q-chip__content col row no-wrap items-center q-anchor--skip"> Reembolsável </div>
    </div>
  </div>
  <div data-v-be666f43="" class="q-card__section q-card__section--vert col-12 col-sm-3">
    <hr data-v-be666f43="" class="q-separator q-separator--vertical bg-grey-3 custom-separator" aria-orientation="vertical">
    <p data-v-be666f43="" class="q-mb-xs text-grey-6 text-caption"> A partir de: </p>
    <p data-v-be666f43="" class="q-mb-xs"> R$ <span data-v-be666f43="" class="text-h5">100,00</span></p>
    <p data-v-be666f43="" class="q-mb-none text-grey-6 text-caption"> R$33,33/noite </p>
    <p data-v-be666f43="" class="text-grey-7 q-mb-xl text-caption"> Impostos inclusos </p>
    <div data-v-be666f43="" class="row"><button data-v-be666f43="" class="q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--rectangle q-btn--rounded bg-primary text-white q-btn--actionable q-focusable q-hoverable q-btn--no-uppercase col-12 col-sm-auto" style="font-size: 14px; min-width: 140px;" tabindex="0" type="button"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"> Selecionar </span></button></div>
  </div>
</div>"
`;
